board/imgtec/pistachio_bub/pistachio.c

```diff
@@ -109,22 +111,42 @@ static const char *get_dtb_macaddr(u32 ifno)
 	if (mac && is_valid_ethaddr((u8 *)mac))
 		return mac;
 
-        return NULL;
+	return NULL;
 }
 #endif
 
 int board_eth_init(bd_t *bs)
 {
+	u_char mac_addr[MAC_ADDR_LEN];
+
 	mfio_setup_ethernet();
 
-	/* try to get a valid macaddr from dtb */
+	/* Order of precedence:
+	 * 1. Check for existing ethaddr environment variable
+	 * 2. Read from OTP
+	 * 3. Fallback on dtb
+	 */
+	memset(mac_addr, 0, MAC_ADDR_LEN);
+	eth_getenv_enetaddr("ethaddr", mac_addr);
+
+#ifdef CONFIG_WINBOND_OTP
+	if (!is_valid_ethaddr(mac_addr)
+	&&  !check_otp_version(VERSION_REG0_OFFSET)) {
+		if (!read_otp_data(ETH_MAC_ADDRESS_OFFSET, MAC_ADDR_LEN, (char *)mac_addr)
+		&&  is_valid_ethaddr(mac_addr))
+			eth_setenv_enetaddr("ethaddr", (u8 *)mac_addr);
+		else
+			printf("Could not read MAC address from OTP\n");
+	}
+#endif
 #ifdef CONFIG_OF_CONTROL
```

Comment:
Enabling CONFIG_OF_CONTROL means that we will retrieve the MAC address from the DTB if we did not find a valid MAC in the OTP.